File: model/src/main/java/ru/it/sd/model/Problem.java
package ru.it.sd.model;

import ru.it.sd.meta.ClassMeta;
import ru.it.sd.meta.FieldMeta;

import java.util.Date;

/**
 * Модельный класс для "Проблем"
 * @author nsychev
 * @since 05.06.2018
 */
@ClassMeta(tableName = "itsm_problems", tableAlias = "pro")
public class Problem implements HasId, HasStatus, HasFolder, HasAssignment{
    /** ID */
    @FieldMeta(columnName = "pro_oid", key = true)
    private Long id;
    /** Номер */
    @FieldMeta(columnName = "pro_id")
    private Long no;
    /** Статус */
    @FieldMeta(columnName = "pro_sta_oid")
    private EntityStatus status;
    /** Инициатор */
    @FieldMeta(columnName = "pro_requestor_per_oid")
    private Person initiator;
    /** Объект обслуживания */
    @FieldMeta(columnName = "pro_cit_oid")
    private ConfigurationItem configurationItem;
    /** Система */
    //todo EntityCode4
    /** Тема */
    @FieldMeta(columnName = "pro_description")
    private String subject;
    /** Информация */
    @FieldMeta(columnName = "pri_information", tableAlias = "pri")
    private String description;
    /** Ссылки на лог-файлы */
    @FieldMeta(columnName = "pcf_problemtext4", tableAlias = "pcf")
    private String logLinks;
    /** Ссылка в jira */
    @FieldMeta(columnName = "pcf_problemtext1", tableAlias = "pcf")
    private String jiraLink;
    /** Номер/дата обращения к вендору */
    @FieldMeta(columnName = "pcf_problemtext3", tableAlias = "pcf")
    private String toVendor;
    /** Обходное решение */
    @FieldMeta(columnName = "pr2_4k2")
    private String workaround;
    /** Решение */
    @FieldMeta(columnName = "prs_solution")
    private String solution;
    /** Комментарий инициатору */
    @FieldMeta(columnName = "pro_workaround")
    private String commentToInitiator;
    /** Приоритет */
    @FieldMeta(columnName = "pro_imp_oid")
    private EntityPriority priority;
    /** Крайний срок */
    @FieldMeta(columnName = "pro_deadline")
    private Date deadline;
    /** Выполнено(дата) */
    @FieldMeta(columnName = "pro_actualfinish")
    private Date resolvedDate;
    /** Закрыто(дата) */
    @FieldMeta(columnName = "pro_latefinish")
    private Date closureDate;
    /** Просрочена*/
    @FieldMeta(columnName = "pcf_boolean12")
    private Boolean isOverdue;
    /** Кем просрочена*/
    @FieldMeta(columnName = "pcf_proshorttext2")
    private String whoOverdue;
    /** План завершения*/
    @FieldMeta(columnName = "pro_planfinish")
    private Date planFinish;
    /** Причина переноса крайнего срока*/
    @FieldMeta(columnName = "pcf_problemtext2")
    private String deferralReason;
    /** Назначено */
    private Assignment assignment;
    @FieldMeta(columnName = "pro_cat_oid")
    private EntityCategory category;
    @FieldMeta(columnName = "pro_cla_oid")
    private EntityClassification classification;

    //todo функционал
    @FieldMeta(columnName = "pro_clo_oid")
    private EntityClosureCode closureCode;
    /** Папка */
    @FieldMeta(columnName = "pro_poo_oid")
    private Folder folder;
    /** Не включать в отчет Заказчику*/
    @FieldMeta(columnName = "pcf_boolean1")
    private Boolean notAttachInReport;

    //todo проявляется в версии
    //todo решено в версии
    /** Дата выпуска версии*/
    @FieldMeta(columnName = "pcf_problemdate2")
    private Date versionDate;
    /** Комментарий исполнителю*/
    @FieldMeta(columnName = "pr1_4k1")
    private String commentToExecutor;

    @Override
    public Assignment getAssignment() {
        return assignment;
    }

    @Override
    public void setAssignment(Assignment assignment) {
        this.assignment = assignment;
    }

    @Override
    public Folder getFolder() {
        return folder;
    }

    @Override
    public void setFolder(Folder folder) {
        this.folder = folder;
    }

    @Override
    public Long getId() {
        return id;
    }

    @Override
    public void setId(Long id) {
        this.id = id;
    }

    @Override
    public EntityStatus getStatus() {
        return status;
    }

    @Override
    public void setStatus(EntityStatus status) {
        this.status = status;
    }

    public Long getNo() {
        return no;
    }

    public void setNo(Long no) {
        this.no = no;
    }

    public Person getInitiator() {
        return initiator;
    }

    public void setInitiator(Person initiator) {
        this.initiator = initiator;
    }

    public ConfigurationItem getConfigurationItem() {
        return configurationItem;
    }

    public void setConfigurationItem(ConfigurationItem configurationItem) {
        this.configurationItem = configurationItem;
    }

    public String getSubject() {
        return subject;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getLogLinks() {
        return logLinks;
    }

    public void setLogLinks(String logLinks) {
        this.logLinks = logLinks;
    }

    public String getJiraLink() {
        return jiraLink;
    }

    public void setJiraLink(String jiraLink) {
        this.jiraLink = jiraLink;
    }

    public String getToVendor() {
        return toVendor;
    }

    public void setToVendor(String toVendor) {
        this.toVendor = toVendor;
    }

    public String getWorkaround() {
        return workaround;
    }

    public void setWorkaround(String workaround) {
        this.workaround = workaround;
    }

    public String getSolution() {
        return solution;
    }

    public void setSolution(String solution) {
        this.solution = solution;
    }

    public String getCommentToInitiator() {
        return commentToInitiator;
    }

    public void setCommentToInitiator(String commentToInitiator) {
        this.commentToInitiator = commentToInitiator;
    }

    public EntityPriority getPriority() {
        return priority;
    }

    public void setPriority(EntityPriority priority) {
        this.priority = priority;
    }

    public Date getDeadline() {
        return deadline;
    }

    public void setDeadline(Date deadline) {
        this.deadline = deadline;
    }

    public Date getResolvedDate() {
        return resolvedDate;
    }

    public void setResolvedDate(Date resolvedDate) {
        this.resolvedDate = resolvedDate;
    }

    public Date getClosureDate() {
        return closureDate;
    }

    public void setClosureDate(Date closureDate) {
        this.closureDate = closureDate;
    }

    public Boolean getOverdue() {
        return isOverdue;
    }

    public void setOverdue(Boolean overdue) {
        isOverdue = overdue;
    }

    public String getWhoOverdue() {
        return whoOverdue;
    }

    public void setWhoOverdue(String whoOverdue) {
        this.whoOverdue = whoOverdue;
    }

    public Date getPlanFinish() {
        return planFinish;
    }

    public void setPlanFinish(Date planFinish) {
        this.planFinish = planFinish;
    }

    public String getDeferralReason() {
        return deferralReason;
    }

    public void setDeferralReason(String deferralReason) {
        this.deferralReason = deferralReason;
    }

    public EntityCategory getCategory() {
        return category;
    }

    public void setCategory(EntityCategory category) {
        this.category = category;
    }

    public EntityClassification getClassification() {
        return classification;
    }

    public void setClassification(EntityClassification classification) {
        this.classification = classification;
    }

    public EntityClosureCode getClosureCode() {
        return closureCode;
    }

    public void setClosureCode(EntityClosureCode closureCode) {
        this.closureCode = closureCode;
    }

    public Boolean getNotAttachInReport() {
        return notAttachInReport;
    }

    public void setNotAttachInReport(Boolean notAttachInReport) {
        this.notAttachInReport = notAttachInReport;
    }

    public Date getVersionDate() {
        return versionDate;
    }

    public void setVersionDate(Date versionDate) {
        this.versionDate = versionDate;
    }

    public String getCommentToExecutor() {
        return commentToExecutor;
    }

    public void setCommentToExecutor(String commentToExecutor) {
        this.commentToExecutor = commentToExecutor;
    }
}

File: src/api/entity/problem.ts
import {EntityTypes} from './util/entity-types';

/**
 * Сущность - "Проблема"
 * @class
 * @name Problem
 * @mixes ENTITY_MIXIN.Historyable
 * @mixes ENTITY_MIXIN.Approvable
 * @mixes ENTITY_MIXIN.AttachmentsHolder
 * @mixes ENTITY_MIXIN.Accessible
 * @extends EditableEntity
 */
class Problem {
    static readonly entityTypeId:EntityTypes = EntityTypes.Problem;
    /**
     * Номер
     * @property
     * @name Problem#no
     * @type {number}
     */
     no: number;

    /**
     * Статус
     * @property
     * @name Problem#status
     * @type {EntityStatus}
     */
        status;

    /**
     * Инициатор
     * @property
     * @name Problem#initiator
     * @type {Person}
     */
     initiator;

    /**
     * Объект обслуживания
     * @property
     * @name Problem#configurationItem
     * @type {ConfigurationItem}
     */
     configurationItem;

    /**
     * Тема
     * @property
     * @name Problem#subject
     * @type {string}
     */
     subject: string;

    /**
     * Описание
     * @property
     * @name Problem#description
     * @type {string}
     */
     description: string;

    /**
     * Ссылки на логи
     * @property
     * @name Problem#logLinks
     * @type {string}
     */
     logLinks: string;

    /**
     * Ссылка на пробелму в jira
     * @property
     * @name Problem#jiraLink
     * @type {string}
     */
     jiraLink: string;
    /**
     * @property
     * @name Problem#toVendor
     * @type {string}
     */
     toVendor: string;
    /**
     * Обходное решение
     * @property
     * @name Problem#workaround
     * @type {string}
     */
     workaround: string;

    /**
     * Решение
     * @property
     * @name Problem#solution
     * @type {string}
     */
     solution: string;

    /**
     * Приоритет
     * @property
     * @name Problem#priority
     * @type {EntityPriority}
     */
        priority;

    /**
     * Крайний срок
     * @property
     * @name Problem#deadline
     * @type {Date}
     */
     deadline: Date;

    /**
     * Дата фактического выполнения
     * @property
     * @name Problem#resolvedDate
     * @type {Date}
     */
     resolvedDate: Date;

    /**
     * Дата закрытия
     * @property
     * @name Problem#closureDate
     * @type {Date}
     */
     closureDate: Date;

    /**
     * Проблема просрочена
     * @property
     * @name Problem#isOverdue
     * @type {Date}
     */
     isOverdue: Date;

    /**
     * Персона, что просрочила проблему
     * @property
     * @name Problem#whoOverdue
     * @type {Person}
     */
     whoOverdue;

    /**
     * План окночания
     * @property
     * @name Problem#planFinish
     * @type {Date}
     */
     planFinish: Date;

    /**
     * Причина отсрочки
     * @property
     * @name Problem#deferralReason
     * @type {string}
     */
     deferralReason: string;

    /**
     * Сущность "назначено"
     * @property
     * @name Problem#assignment
     * @type {EntityAssignment}
     */
        assignment;

    /**
     * Категория
     * @property
     * @name Problem#category
     * @type {EntityCategory}
     */
     category;

    /**
     * Классификация
     * @property
     * @name Problem#classification
     * @type {EntityClassification}
     */
     classification;

    /**
     * Код завершения
     * @property
     * @name Problem#closureCode
     * @type {EntityClosureCode}
     */
     closureCode;

    /**
     * Папка
     * @property
     * @name Problem#folder
     * @type {Folder}
     */
     folder;

    /**
     * Не включать в отчет заказчику
     * @property
     * @name Problem#isOverdue
     * @type {Date}
     */
     notAttachInReport: Date;

    /**
     *
     * @property
     * @name Problem#versionDate
     * @type {Date}
     */
     versionDate: Date;

    toString() {
        return String(this.no);
    }

}
